### PR TITLE
[JENKINS-21905] Present human readable error when archiving unreadable artifacts

### PR DESCRIPTION
**Pull request by olivergondza**
[JENKINS-21905](https://issues.jenkins-ci.org/browse/JENKINS-21905)

One of the reocuring reason for the cryptic failure in the JIRA is the fact artifacts are not readable for user running slave agent. The exceptions got a bit better since then so it involves the filename and the fact permissions was denied, but user is still staring at ~70 lines of stacktraces. Consolidate this to an easy to grasp message.

To do this, several levels of nested exceptions was removed as it was effectively hiding the root cause.

### Proposed changelog entries

* JENKINS-21905 - Diagnose failure to archive artifacts with insufficient permissions.
  * Related issue: https://issues.jenkins-ci.org/browse/JENKINS-19473

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### Error

Before:

```
Archiving artifacts
ERROR: Failed to archive artifacts: myfile
java.nio.file.AccessDeniedException: /some/path/workspace/test0/myfile
  at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
  at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
  at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
  at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
  at java.nio.file.Files.newByteChannel(Files.java:361)
  at java.nio.file.Files.newByteChannel(Files.java:407)
  at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
  at java.nio.file.Files.newInputStream(Files.java:152)
  at hudson.util.io.TarArchiver.visit(TarArchiver.java:106)
  at hudson.util.DirScanner.scanSingle(DirScanner.java:49)
  at hudson.FilePath$ExplicitlySpecifiedDirScanner.scan(FilePath.java:2823)
  at hudson.FilePath.writeToTar(FilePath.java:2269)
  at hudson.FilePath.access$2100(FilePath.java:197)
  at hudson.FilePath$45.invoke(FilePath.java:2212)
  at hudson.FilePath$45.invoke(FilePath.java:2208)
  at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2750)
  at hudson.remoting.UserRequest.perform(UserRequest.java:181)
  at hudson.remoting.UserRequest.perform(UserRequest.java:52)
  at hudson.remoting.Request$2.run(Request.java:336)
  at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
Caused: java.io.IOException: Error writing to tar file from: /some/path/workspace/test0/myfile
  at hudson.util.io.TarArchiver.visit(TarArchiver.java:113)
  at hudson.util.DirScanner.scanSingle(DirScanner.java:49)
  at hudson.FilePath$ExplicitlySpecifiedDirScanner.scan(FilePath.java:2823)
  at hudson.FilePath.writeToTar(FilePath.java:2269)
  at hudson.FilePath.access$2100(FilePath.java:197)
  at hudson.FilePath$45.invoke(FilePath.java:2212)
  at hudson.FilePath$45.invoke(FilePath.java:2208)
  at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2750)
  at hudson.remoting.UserRequest.perform(UserRequest.java:181)
  at hudson.remoting.UserRequest.perform(UserRequest.java:52)
  at hudson.remoting.Request$2.run(Request.java:336)
  at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
  at ......remote call to slave0(Native Method)
  at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1554)
  at hudson.remoting.UserResponse.retrieve(UserRequest.java:281)
  at hudson.remoting.Channel$2.adapt(Channel.java:869)
Caused: java.util.concurrent.ExecutionException
  at hudson.remoting.Channel$2.adapt(Channel.java:871)
  at hudson.remoting.Channel$2.adapt(Channel.java:866)
  at hudson.remoting.FutureAdapter.get(FutureAdapter.java:55)
  at hudson.FilePath.copyRecursiveTo(FilePath.java:2231)
Caused: java.io.IOException
  at hudson.FilePath.copyRecursiveTo(FilePath.java:2233)
  at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:61)
  at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:244)
  at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
  at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
  at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:730)
  at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:676)
  at hudson.model.Build$BuildExecution.post2(Build.java:186)
  at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:621)
  at hudson.model.Run.execute(Run.java:1760)
  at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
  at hudson.model.ResourceController.execute(ResourceController.java:97)
  at hudson.model.Executor.run(Executor.java:415)
Build step 'Archive the artifacts' changed build result to FAILURE
Finished: FAILURE
```

After:
```
ERROR: Step ‘Archive the artifacts’ failed: java.nio.file.AccessDeniedException: /some/path/workspace/test0/myfile
```

### CODE REVIEW

**Comment by daniel-beck** [comment]
Would it not be useful to print the stack on a low (e.g. FINE) log level to at least give a chance to figure out more details in case this catches more than it should?

**Comment by olivergondza** [comment]
Sure, this is how it would look like:

```
WARNING: Diagnosing anticipated Exception
java.nio.file.AccessDeniedException: /home/ogondza/code/jenkins/jenkins/test/target/jenkinsTests.tmp/jenkins8657221315180995243test/workspace/test0/myfile
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at hudson.util.io.TarArchiver.visit(TarArchiver.java:106)
	at hudson.util.DirScanner.scanSingle(DirScanner.java:49)
	at hudson.FilePath$ExplicitlySpecifiedDirScanner.scan(FilePath.java:2834)
	at hudson.FilePath.writeToTar(FilePath.java:2280)
	at hudson.FilePath.access$2100(FilePath.java:197)
	at hudson.FilePath$45.invoke(FilePath.java:2217)
	at hudson.FilePath$45.invoke(FilePath.java:2213)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2761)
	at hudson.remoting.UserRequest.perform(UserRequest.java:181)
	at hudson.remoting.UserRequest.perform(UserRequest.java:52)
	at hudson.remoting.Request$2.run(Request.java:336)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at ......remote call to slave0(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1554)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:281)
	at hudson.remoting.Channel$2.adapt(Channel.java:869)
	at hudson.remoting.Channel$2.adapt(Channel.java:866)
	at hudson.remoting.FutureAdapter.get(FutureAdapter.java:55)
	at hudson.FilePath.copyRecursiveTo(FilePath.java:2237)
	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:61)
	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:245)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:730)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:676)
	at hudson.model.Build$BuildExecution.post2(Build.java:186)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:621)
	at hudson.model.Run.execute(Run.java:1760)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:415)
```

**Comment by daniel-beck** [comment]
@olivergondza Thanks!

**Comment by olivergondza** [comment]
Skipping the test on windows after all as the permission manipulation does not seem to work there.

**Comment by oleg-nenashev** [comment]
I will try to review it next week. Retriggering CI

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -2199,7 +2199,12 @@ public Void invoke(File f, VirtualChannel channel) throws IOException {
                 future.get();
                 return future2.get();
             } catch (ExecutionException e) {
-                throw new IOException(e);
+                Throwable cause = e.getCause();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]: It loses the track of `ExecutionException`. I would add it to suppressed at least

File: core/src/main/java/hudson/util/io/TarArchiver.java [hunk 2]
```diff
@@ -102,16 +102,20 @@ public void visit(File file, String relativePath) throws IOException {
...
             if (!file.isDirectory()) {
                 // ensure we don't write more bytes than the declared when we created the entry
-                
+
                 try (InputStream fin = Files.newInputStream(file.toPath());
                      BoundedInputStream in = new BoundedInputStream(fin, size)) {
-                    int len;
-                    while ((len = in.read(buf)) >= 0) {
-                        tar.write(buf, 0, len);
+                    // Separate tray block not to wrap exception thrown while opening the input stream into an exception
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/io/TarArchiver.java]: s/tray/try

**Review by oleg-nenashev — COMMENTED** [review]
The changes look good to me, excepting FilePath. Something is missing there imho

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (olivergondza) [reply core/src/main/java/hudson/FilePath.java]: That is intentional as that is just a wrapper to clearly signify what exceptions are coming from the asynchronous framework and which are thrown by the task itself. We know it is the task that failed and I do not see much value in informing people it was executed asynchronously. The main motivation though was to get rid of the clutter there was a lot of.

**Comment by olivergondza** [comment]
Review comments addressed.

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/FilePath.java]: Hmm, OK. Maybe it would be useful to push more details, e.g. if the FINE logging enabled for the class.

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers any votes?

**Comment by oleg-nenashev** [comment]
@olivergondza It's definitely ready to go, but I am not sure how to properly note it and how to properly squash it since there are 2 issues mentioned in commit history. IIUC JENKINS-21905 is just a diagnosability improvement on the top of JENKINS-19473.

Could you please integrate the change in a way you find the best?

**Comment by olivergondza** [comment]
Rebased and retagged to refer to JENKINS-21905 I intend to close so all future reports will be relevant for eventual finetuning.

**Comment by daniel-beck** [comment]
> To do this, several levels of nested exceptions was removed as it was effectively hiding the root cause.

Notably since https://jenkins.io/changelog/#v2.43 we're already listing the innermost exception on top. The actual error here was already on the first 2-3 lines.

**Comment by olivergondza** [comment]
@daniel-beck, right, for user consumption, yes. Though, the exception was cumbersome to act upon when wrapped in all that clutter and that is what this PR took advantage of.

---

Not that I agree we rearrange exception order like this in general - especially if we fail to be consistent doing that. The remoting exceptions are quire hard to navigate on its own and adding potential inverted listing, depending on which part of Jenkins has printed it, is not helping.

**Pull request merged**